File: CS229/Project1/src/ImageUtil.java
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Utility methods for reading and writing a CS229 image.
 * 
 * @author Brian Reber
 */
public class ImageUtil {

	/**
	 * Read a CS229 image from the given file.
	 * 
	 * @param f
	 * The file to read from
	 * @return
	 * The CS229Image representing the image
	 * @throws IOException 
	 */
	public static CS229Image readInCS229Image(File f) throws IOException {
		CS229Image image = new CS229Image();
		InputStream input;
		int tempByte;
		int tempSize;

		input = new FileInputStream(f);
		
		// Read Magic Number and check
		if (input.read() != 0x42) {
			throw new IOException("Not a CS229 Image");
		}
		
		tempByte = input.read();
		
		if (tempByte == 0x01 || tempByte == 0x02 || tempByte == 0x04) {
			image.setChannelSizeR(tempByte);
		}
		
		tempByte = input.read();
		
		if (tempByte == 0x01 || tempByte == 0x02 || tempByte == 0x04) {
			image.setChannelSizeG(tempByte);
		}
		
		tempByte = input.read();
		
		if (tempByte == 0x01 || tempByte == 0x02 || tempByte == 0x04) {
			image.setChannelSizeB(tempByte);
		}
		
		tempByte = input.read();
		
		if (tempByte == 0x00 || tempByte == 0xFF) {
			image.setBW(tempByte == 0x00);
		}
		
		tempSize = input.read();
		tempSize |= (input.read() << 8);
		tempSize |= (input.read() << 8);
		tempSize |= (input.read() << 8);
		
		if (tempSize > 0) {
			image.setWidth(tempSize);
		}
		
		tempSize = input.read();
		tempSize |= (input.read() << 8);
		tempSize |= (input.read() << 8);
		tempSize |= (input.read() << 8);
		
		if (tempSize > 0) {
			image.setHeight(tempSize);
		}
		
		readImageData(input, image);
		
		return image;
	}
	
	/**
	 * Reads the image pixel data from the input stream to the given image data
	 * 
	 * @param input
	 * The input stream containing the data 
	 * @param image
	 * The image to put the pixel data into
	 * @throws IOException
	 */
	private static void readImageData(InputStream input, CS229Image image) throws IOException {
		CS229Image.Pixel[] data = new CS229Image.Pixel[image.getWidth() * image.getHeight() + 2];
		for (int i = 0; i < data.length; i++) {
			data[i] = new CS229Image.Pixel();
		}
		int[] currentBit = new int[3];
		int currentChannel = 0;
		int pos = 0;
		
		if (image.isBW()) {
			/* 
			 * While we haven't finished the file, read in bit by bit
			 * the values of the image. 
			 */
			while ((currentBit[0] = input.read()) != -1) {
				if (image.getChannelSizeR() == 1) {
					/* 4 bit - we want to read 2 pixels for each bit */
					data[pos++].red = (currentBit[0] >> 4) & 0x0F;
					data[pos++].red = currentBit[0] & 0x0F;
				} else if (image.getChannelSizeR() == 2) {
					/* 8 bit - we want to read 1 pixels for each bit */
					data[pos++].red = currentBit[0];
				} else if (image.getChannelSizeR() == 4 && currentChannel != 1) {
					/* 
					 * 16 bit - we want to read 1/2 of a pixel for each bit 
					 * here we set the current position to the current read 
					 * left shifted by 4 bits.
					 */
					data[pos].red = currentBit[0] << 8;
					currentChannel = 1;
				} else if (image.getChannelSizeR() == 4 && currentChannel == 1) {
					/* 
					 * 16 bit (part 2) - we want to read 1/2 of a pixel for each bit 
					 * here we or the current position to the current position, and  
					 * then increment our position
					 */
					data[pos++].red |= currentBit[0];
					currentChannel = 0;
				}
			}
		} else {
			int[] tempData = new int[image.getWidth() * image.getHeight() * 3];
			int i = 0;
			int tempBit = 0;
			int size;
			
			OffsetReadUpper ret = new OffsetReadUpper();
			ret.offset = 0;
			ret.readUpper = 0xF0;
			
			/* 
			 * While we haven't finished the file, read in bit by bit
			 * the values of the image. Since this is a color image,
			 * we need to handle 3 bits at a time, so we need to keep 
			 * track of the current 'channel'
			 */
			while ((tempBit = input.read()) != -1) {
				tempData[i++] = tempBit;
			}
			size = i;
			i = 0;
			while (ret.offset < size) {
				data[i].red = readPackedBits(tempData, ret, image.getChannelSizeR());
				data[i].green = readPackedBits(tempData, ret, image.getChannelSizeG());
				data[i].blue = readPackedBits(tempData, ret, image.getChannelSizeB());
				i++;
			}
		}
		
		image.setData(data);
	}
	
	/**
	 * A small inner class to hold the offset, and which bits to read
	 * for the reading of CS229 images.
	 * 
	 * @author Brian Reber
	 */
	private static class OffsetReadUpper {
		int offset;
		int readUpper;		
	}
	
	/**
	 * Reads bits from the int array given the channel size, and offset
	 * 
	 * @param readByte
	 * The array of data
	 * @param vals
	 * Object containing the offset and which bits to read
	 * @param channelSize
	 * the size of the current channel to read
	 * @return
	 * the value of the data in the given bits
	 */
	private static int readPackedBits(int[] readByte, OffsetReadUpper vals, int channelSize) {
		int data = 0;
		
		if (channelSize == 1) {
			if (vals.readUpper == 0xF0) {
				data = (readByte[vals.offset] & 0xF0) >> 4;
			vals.readUpper = 0x0F;
			} else if (vals.readUpper == 0x0F) {
				data = readByte[vals.offset] & 0x0F;
				vals.readUpper = 0xF0;
				vals.offset++;
			}
		} else if (channelSize == 2) {
			if (vals.readUpper == 0xF0) {
				data = readByte[vals.offset];
				vals.offset++;
				vals.readUpper = 0xF0;
			} else if (vals.readUpper == 0x0F) {
				data = readByte[vals.offset] & 0x0F;
				vals.offset++;
				data |= ((readByte[vals.offset] & 0xF0) >> 4);
				vals.readUpper = 0x0F;
			}
		} else {
			if (vals.readUpper == 0xF0) {
				data = (readByte[vals.offset] << 8);
				vals.offset++;
				data |= readByte[vals.offset];
				vals.offset++;
				vals.readUpper = 0xF0;
			} else if (vals.readUpper == 0x0F) {
				data = ((readByte[vals.offset] & 0x0F) << 12);
				vals.offset++;
				data |= (readByte[vals.offset] << 4);
				vals.offset++;
				data |= ((readByte[vals.offset] & 0xF0) >> 4);
				vals.readUpper = 0x0F;
			}
		}

		return data;
	}
	
	/**
	 * Writes a CS229 image to the given file
	 * 
	 * @param image
	 * The image to write
	 * @param file
	 * The file to write to
	 * @throws IOException 
	 */
	public static void writeCS229Image(CS229Image image, File file) throws IOException {
		OutputStream output;
		output = new FileOutputStream(file);
		
		output.write(0x42);
		
		/* Channel Size */
		output.write(image.getChannelSizeR());
		output.write(image.getChannelSizeG());
		output.write(image.getChannelSizeB());

		/* Black and White */
		output.write(image.isBW() ? 0x00 : 0xFF);
		
		/* Image Width */
		
		output.write((image.getWidth() & 0xFF));
		output.write(((image.getWidth() >> 8) & 0xFF));
		output.write(((image.getWidth() >> 16) & 0xFF));
		output.write(((image.getWidth() >> 24) & 0xFF));
		
		/* Image Height */
		output.write((image.getHeight() & 0xFF));
		output.write(((image.getHeight() >> 8) & 0xFF));
		output.write(((image.getHeight() >> 16) & 0xFF));
		output.write(((image.getHeight() >> 24) & 0xFF));
		
		writeImageData(output, image);
	}
	
	
	/**
	 * Packs bytes correctly given the current byte buffer, and the given data.
	 *
	 * @param  file - the OutputStream to write to
	 * @param  byteToWrite - the buffer of bits that is to be written - this method will modify
	 *				this if when the data is packed it doesn't contain a full byte to write
	 * @param  numBitsInByteToWrite - the number of bits that have been added to the buffer
	 * @param  channelSize - the channel size for the current data
	 * @param  data - the data that needs to be written
	 * @throws IOException 
	 */
	private static void packBitsAndWrite(OutputStream file, Byte byteToWrite, Integer numBitsInByteToWrite, int channelSize, int data) throws IOException {
		if (channelSize == 1) {
			if (numBitsInByteToWrite == 0) {
				/* 
				 * The current channel is 4 bits, and there isn't anything in the buffer 
				 * We will add the data to the buffer
				 */
				byteToWrite = (byte) data;
				numBitsInByteToWrite = 4;
			} else if (numBitsInByteToWrite == 4) {
				/* 
				 * The current channel is 4 bits, and there are 4 bits in the buffer.
				 * We will left-shift the buffer by 4 bits, and or the data to the buffer,
				 * write the data to the file, and clear the buffer
				 */
				byteToWrite = (byte) ((byteToWrite << 4) | data);
				file.write(byteToWrite);
				numBitsInByteToWrite = 0;
				byteToWrite = 0;
			}
		} else if (channelSize == 2) {
			if (numBitsInByteToWrite == 0) {
				/* 
				 * The current channel is 8 bits, and there are 0 bits in the buffer.
				 * We will write the byte size data to the file, and clear the buffer
				 * for good measure
				 */
				file.write(data);
				numBitsInByteToWrite = 0;
				byteToWrite = 0;
			} else if (numBitsInByteToWrite == 4) {
				/* 
				 * The current channel is 8 bits, and there are 4 bits in the buffer.
				 * We will left-shift the buffer by 4 bits, and or the upper 4 bits of data to the buffer,
				 * write the buffer to the file, and then add the lower 4 bits to the buffer
				 */
				byteToWrite = (byte) ((byteToWrite << 4) | ((data & 0xF0) >> 4));
				file.write(byteToWrite);
				numBitsInByteToWrite = 4;
				byteToWrite = (byte) (data & 0x0F);
			}
		} else {
			if (numBitsInByteToWrite == 0) {
				/* 
				 * The current channel is 16 bits, and there are 0 bits in the buffer.
				 * We will write the 2 bytes to the file, and clear the buffer for good measure
				 */
				file.write(data & 0x00FF);
				file.write((data & 0xFF00) >> 8);
				numBitsInByteToWrite = 0;
				byteToWrite = 0;
			} else if (numBitsInByteToWrite == 4) {
				/* 
				 * The current channel is 16 bits, and there are 4 bits in the buffer.
				 * We will left-shift the buffer by 4 bits, and or the upper-most 4 bits of data to the buffer,
				 * write the data to the file.  Then we will write the middle 2 bits to the file,
				 * and finally add the lowest 4 bits to the buffer.
				 */
				byteToWrite = (byte) ((byteToWrite << 4) | ((data & 0xF000) >> 12));
				file.write(byteToWrite);
				byteToWrite = (byte) ((data & 0x0FF0) >> 4);
				file.write(byteToWrite);
				byteToWrite = (byte) (data & 0x000F);
				numBitsInByteToWrite = 4;
			}
		}
	}
		
	/**
	 * Writes the image (pixel) data to the given file
	 *
	 * @param  file - the OutputStream to write to
	 * @param  image - the image struct containing the data to be written
	 * @throws IOException 
	 */
	private static void writeImageData(OutputStream file, CS229Image image) throws IOException {
		int i;
		Byte byteToWrite = 0;
		Integer numBitsInByteToWrite = 0;
		CS229Image.Pixel[] data = image.getData();
		for (i = 0; i < image.getWidth() * image.getHeight(); i++) {
			if (image.isBW()) {
				packBitsAndWrite(file, byteToWrite, numBitsInByteToWrite, image.getChannelSizeR(), data[i].red);
			} else {
				packBitsAndWrite(file, byteToWrite, numBitsInByteToWrite, image.getChannelSizeR(), data[i].red);
				packBitsAndWrite(file, byteToWrite, numBitsInByteToWrite, image.getChannelSizeG(), data[i].green);
				packBitsAndWrite(file, byteToWrite, numBitsInByteToWrite, image.getChannelSizeB(), data[i].blue);
			}
		}
		
		// If we ended with data still in the byteToWrite buffer, output that, and fil in the rest
		// with zeros
		if (numBitsInByteToWrite == 4) {
			file.write((byteToWrite << 4) & 0xF0);
		}
	}
}

File: CS229/Project1/src/imageutil.c
/**
 * @file imageutil.c
 * @author  Brian Reber
 *
 * @section DESCRIPTION
 *
 * This file contains utility functions for CS229 images
 * including reading image values, and image data.
 */

#include <stdio.h>
#include "imageutil.h"

/**
 * Gets the channel size in terms of 4, 8 or 16.
 *
 * This essentially is a map, mapping the actual number of bits
 * in the channel to the code we use to write it in the image
 *
 * @param  key - represents the hex value we stored in the image
 * @return 4, 8, 16 depending on the channel size. -1 if for some
 *		reason the param doesn't match any of the values
 */
int getChannelSize(unsigned int key) {
	if (key == FOUR_BIT) {
		return 4;
	} else if (key == EIGHT_BIT) {
		return 8;
	} else if (key == SIXTEEN_BIT) {
		return 16;
	}
	
	fprintf(stderr, "Unknown channel size\n");
	return -1;
}

/**
 * Gets the corresponding black constant for the given channel size
 *
 * @param  channelSize - represents the channel size
 * @return 4 bit = 0x0, 8 bit = 0x00, 16 bit = 0x0000, -1 if channel size doesn't
 *		match any of the three options
 */
int getBlack(unsigned int channelSize) {
	switch (channelSize) {
		case FOUR_BIT:
			return BLACK_4;
		case EIGHT_BIT:
			return BLACK_8;
			break;
		case SIXTEEN_BIT:
			return BLACK_16;
			break;
		default:
			return -1;
	}
}

/**
 * Gets the corresponding white constant for the given channel size
 *
 * @param  channelSize - represents the channel size
 * @return 4 bit = 0xF, 8 bit = 0xFF, 16 bit = 0xFFFF, -1 if channel size doesn't
 *		match any of the three options
 */
int getWhite(unsigned int channelSize) {
	switch (channelSize) {
		case FOUR_BIT:
			return WHITE_4;
		case EIGHT_BIT:
			return WHITE_8;
			break;
		case SIXTEEN_BIT:
			return WHITE_16;
			break;
		default:
			return -1;
	}
}

/**
 * Reads the image values (bw, width, height, channel size) from the given CS229 file.
 *
 * @param  *file - the file to read from
 * @param  *image_ptr - the image struct to put the image properties in
 */
int readCSImageValues(FILE *file, struct image *image_ptr) {
	unsigned int is229;
	unsigned int temp;
	
	/* The "Magic" character */
	is229 = fgetc(file);
	if (is229 != CS229) {
		fprintf(stderr, "This isn't a CS229 image\n");
		return -1;
	}
	
	/* Channel Size */
	image_ptr->channelSize.red = fgetc(file);
	image_ptr->channelSize.green = fgetc(file);
	image_ptr->channelSize.blue = fgetc(file);
	
	/* Black and White */
	temp = fgetc(file);
	image_ptr->isBW = (temp == ZERO);
	
	/* 
	 * Getting the height and width back to one number from
	 * the 4 bits involves setting the LSB value to the variable
	 * and then each consecutive (more significant) gets shifted left
	 * 8, and then "OR"ed with what is in the variable currently
	 */
	
	/* Image Width */
	image_ptr->imWidth = fgetc(file);
	temp = fgetc(file);
	image_ptr->imWidth = image_ptr->imWidth | (temp << 8);
	temp = fgetc(file);
	image_ptr->imWidth = image_ptr->imWidth | (temp << 8);
	temp = fgetc(file);
	image_ptr->imWidth = image_ptr->imWidth | (temp << 8);
	
	/* Image Height */
	image_ptr->imHeight = fgetc(file);
	temp = fgetc(file);
	image_ptr->imHeight = image_ptr->imHeight | (temp << 8);
	temp = fgetc(file);
	image_ptr->imHeight = image_ptr->imHeight | (temp << 8);
	temp = fgetc(file);
	image_ptr->imHeight = image_ptr->imHeight | (temp << 8);
	
	return 0;
}

/**
 * Reads the image (pixel) data from the given file
 *
 * @param  *file - the file to read from
 * @param  *image_ptr - the image struct to put the image pixel data into
 */
void readImageData(FILE *file, struct image *image_ptr) {
	unsigned int currentBit[3];
	int currentChannel = 0;
	int pos = 0;
	
	image_ptr->data = (struct pixel *) malloc(image_ptr->imWidth * image_ptr->imHeight * sizeof(struct pixel));
	
	if (image_ptr->data == NULL) {
		fprintf(stderr, "There was an error allocating the necessary memory to hold this image\n");
		return;
	}
	
	if (image_ptr->isBW) {
		/* 
		 * While we haven't finished the file, read in bit by bit
		 * the values of the image. 
		 */
		while ((currentBit[0] = fgetc(file)) != EOF) {
			if (image_ptr->channelSize.red == 1) {
				/* 4 bit - we want to read 2 pixels for each bit */
				image_ptr->data[pos++].red = currentBit[0] & 0xF0;
				image_ptr->data[pos++].red = currentBit[0] & 0x0F; 
			} else if (image_ptr->channelSize.red == 2) {
				/* 8 bit - we want to read 1 pixels for each bit */
				image_ptr->data[pos++].red = currentBit[0];
			} else if (image_ptr->channelSize.red == 4 && currentChannel != 1) {
				/* 
				 * 16 bit - we want to read 1/2 of a pixel for each bit 
				 * here we set the current position to the current read 
				 * left shifted by 4 bits.
				 */
				image_ptr->data[pos].red = currentBit[0] << 8;
				currentChannel = 1;
			} else if (image_ptr->channelSize.red == 4 && currentChannel == 1) {
				/* 
				 * 16 bit (part 2) - we want to read 1/2 of a pixel for each bit 
				 * here we or the current position to the current position, and  
				 * then increment our position
				 */
				image_ptr->data[pos++].red |= currentBit[0];
				currentChannel = 0;
			}
		}
	} else {
		unsigned char *tempData = (unsigned char *) malloc(image_ptr->imWidth * image_ptr->imHeight * sizeof(struct pixel));
		int i = 0;
		unsigned int tempBit = 0;
		int totalBytesRead = 0;
		int offset = 0;
		int readUpper = 0xF0;
		
		/* 
		 * While we haven't finished the file, read in bit by bit
		 * the values of the image. Since this is a color image,
		 * we need to handle 3 bits at a time, so we need to keep 
		 * track of the current 'channel'
		 */
		while ((tempBit = getc(file)) != EOF) {
			tempData[i++] = tempBit;
		}
		
		totalBytesRead = i;
		
		i = 0;
		while ((offset < totalBytesRead) && (i < (image_ptr->imWidth * image_ptr->imHeight))) {
			readPackedBits(tempData, &offset, &readUpper, image_ptr->channelSize.red, &image_ptr->data[i].red);
			readPackedBits(tempData, &offset, &readUpper, image_ptr->channelSize.green, &image_ptr->data[i].green);
			readPackedBits(tempData, &offset, &readUpper, image_ptr->channelSize.blue, &image_ptr->data[i].blue);
			i++;
		}
	}
}

/**
 * Reads bits from the int array given the channel size, and offset
 *
 * @param  *readByte - the array of data
 * @param  *offset - where we are in the data array
 * @param  *readUpper - the mask to use for the current byte to get the necessary bits
 * @param  channelSize - the channel size for the current data
 * @param  *data - the place to put the read data
 */
void readPackedBits(unsigned char *readByte, int *offset, int *readUpper, unsigned char channelSize, unsigned short int *data) {
	if (channelSize == 1) {
		if (*readUpper == 0xF0) {
			*data = (readByte[*offset] & 0xF0) >> 4;
			*readUpper = 0x0F;
		} else if (*readUpper == 0x0F) {
			*data = readByte[*offset] & 0x0F;
			*readUpper = 0xF0;
			(*offset)++;
		}
	} else if (channelSize == 2) {
		if (*readUpper == 0xF0) {
			*data = readByte[*offset];
			(*offset)++;
			*readUpper = 0xF0;
		} else if (*readUpper == 0x0F) {
			*data = readByte[*offset] & 0x0F;
			(*offset)++;
			*data |= ((readByte[*offset] & 0xF0) >> 4);
			*readUpper = 0x0F;
		}
	} else {
		if (*readUpper == 0xF0) {
			*data = (readByte[*offset] << 8);
			(*offset)++;
			*data |= readByte[*offset];
			(*offset)++;
			*readUpper = 0xF0;
		} else if (*readUpper == 0x0F) {
			*data = ((readByte[*offset] & 0x0F) << 12);
			(*offset)++;
			*data |= (readByte[*offset] << 4);
			(*offset)++;
			*data |= ((readByte[*offset] & 0xF0) >> 4);
			*readUpper = 0x0F;
		}
	}
}

/**
 * Packs bytes correctly given the current byte buffer, and the given data.
 *
 * @param  *file - the file to write to
 * @param  *byteToWrite - the buffer of bits that is to be written - this function will modify
 *				this if when the data is packed it doesn't contain a full byte to write
 * @param  *numBitsInByteToWrite - the number of bits that have been added to the buffer
 * @param  channelSize - the channel size for the current data
 * @param  data - the data that needs to be written
 */
void packBitsAndWrite(FILE *file, unsigned char *byteToWrite, int *numBitsInByteToWrite, unsigned char channelSize, unsigned short int data) {
	if (channelSize == 1) {
		if (*numBitsInByteToWrite == 0) {
			/* 
			 * The current channel is 4 bits, and there isn't anything in the buffer 
			 * We will add the data to the buffer
			 */
			*byteToWrite = data;
			*numBitsInByteToWrite = 4;
		} else if (*numBitsInByteToWrite == 4) {
			/* 
			 * The current channel is 4 bits, and there are 4 bits in the buffer.
			 * We will left-shift the buffer by 4 bits, and or the data to the buffer,
			 * write the data to the file, and clear the buffer
			 */
			*byteToWrite = (*byteToWrite << 4) | data;
			putc(*byteToWrite, file);
			*numBitsInByteToWrite = 0;
			*byteToWrite = 0;
		}
	} else if (channelSize == 2) {
		if (*numBitsInByteToWrite == 0) {
			/* 
			 * The current channel is 8 bits, and there are 0 bits in the buffer.
			 * We will write the byte size data to the file, and clear the buffer
			 * for good measure
			 */
			putc(data, file);
			*numBitsInByteToWrite = 0;
			*byteToWrite = 0;
		} else if (*numBitsInByteToWrite == 4) {
			/* 
			 * The current channel is 8 bits, and there are 4 bits in the buffer.
			 * We will left-shift the buffer by 4 bits, and or the upper 4 bits of data to the buffer,
			 * write the buffer to the file, and then add the lower 4 bits to the buffer
			 */
			*byteToWrite = (*byteToWrite << 4) | ((data & 0xF0) >> 4);
			putc(*byteToWrite, file);
			*numBitsInByteToWrite = 4;
			*byteToWrite = data & 0x0F;
		}
	} else {
		if (*numBitsInByteToWrite == 0) {
			/* 
			 * The current channel is 16 bits, and there are 0 bits in the buffer.
			 * We will write the 2 bytes to the file, and clear the buffer for good measure
			 */
			putc(data & 0x00FF, file);
			putc((data & 0xFF00) >> 8, file);
			*numBitsInByteToWrite = 0;
			*byteToWrite = 0;
		} else if (*numBitsInByteToWrite == 4) {
			/* 
			 * The current channel is 16 bits, and there are 4 bits in the buffer.
			 * We will left-shift the buffer by 4 bits, and or the upper-most 4 bits of data to the buffer,
			 * write the data to the file.  Then we will write the middle 2 bits to the file,
			 * and finally add the lowest 4 bits to the buffer.
			 */
			*byteToWrite = (*byteToWrite << 4) | ((data & 0xF000) >> 12);
			putc(*byteToWrite, file);
			*byteToWrite = (data & 0x0FF0) >> 4;
			putc(*byteToWrite, file);
			*byteToWrite = (data & 0x000F);
			*numBitsInByteToWrite = 4;
		}
	}
}

/**
 * Writes the image (pixel) data to the given file
 *
 * @param  *file - the file to write to
 * @param  *image_ptr - the image struct containing the data to be written
 */
void writeImageData(FILE *file, struct image *image_ptr) {
	int i;
	unsigned char byteToWrite;
	int numBitsInByteToWrite = 0;
	
	for (i = 0; i < image_ptr->imWidth * image_ptr->imHeight; i++) {
		if (image_ptr->isBW) {
			packBitsAndWrite(file, &byteToWrite, &numBitsInByteToWrite, image_ptr->channelSize.red, image_ptr->data[i].red);
		} else {
			packBitsAndWrite(file, &byteToWrite, &numBitsInByteToWrite, image_ptr->channelSize.red, image_ptr->data[i].red);
			packBitsAndWrite(file, &byteToWrite, &numBitsInByteToWrite, image_ptr->channelSize.green, image_ptr->data[i].green);
			packBitsAndWrite(file, &byteToWrite, &numBitsInByteToWrite, image_ptr->channelSize.blue, image_ptr->data[i].blue);
		}
	}
	
	/*
	 * If we ended with data still in the byteToWrite buffer, output that, and fil in the rest
	 * with zeros
	 */
	if (numBitsInByteToWrite == 4) {
		putc((byteToWrite << 4) & 0xF0, file);
	}
}

/**
 * Writes a CS229 image to the given file
 *
 * @param  *file - the file to write to
 * @param  *image_ptr - the image struct containing the data to be written to CS
 */
void writeCSImage(FILE *file, struct image *image_ptr) {
	/* The "Magic" character */
	putc(CS229, file);
	
	/* Channel Size */
	putc(image_ptr->channelSize.red, file);
	putc(image_ptr->channelSize.green, file);
	putc(image_ptr->channelSize.blue, file);

	/* Black and White */
	putc((image_ptr->isBW) ? 0x00 : 0xFF, file);
	
	/* Image Width */
	putc((image_ptr->imWidth & 0xFF), file);
	putc(((image_ptr->imWidth >> 8) & 0xFF), file);
	putc(((image_ptr->imWidth >> 16) & 0xFF), file);
	putc(((image_ptr->imWidth >> 24) & 0xFF), file);
	
	/* Image Height */
	putc((image_ptr->imHeight & 0xFF), file);
	putc(((image_ptr->imHeight >> 8) & 0xFF), file);
	putc(((image_ptr->imHeight >> 16) & 0xFF), file);
	putc(((image_ptr->imHeight >> 24) & 0xFF), file);
	
	writeImageData(file, image_ptr);
}

/**
 * Writes the PPM image to stdout
 *
 * @param  *image_ptr - the image struct containing the data to be written to ppm
 */
void writePPMImage(struct image *image_ptr) {
	if (image_ptr->isBW) {
		fprintf(stderr, "PPM does not support BW images\n");
		return;
	}
	
	/* The "Magic" character */
	fprintf(stdout, PPM_MAGIC);
	
	fprintf(stdout, "\n");
	
	/* Height and Width */
	fprintf(stdout, "%d %d\n", image_ptr->imWidth, image_ptr->imHeight); 
	
	/* Maxval */
	fprintf(stdout, "%d\n", (image_ptr->channelSize.red == EIGHT_BIT) ? WHITE_8 : WHITE_16);
	
	writeImageData(stdout, image_ptr);
}
